### PR TITLE
audio: Add headset example

### PR DESCRIPTION
**Pull request by kasjer**
**Describe the PR**
This example code creates USB Audio 2.0 headset device.
Device has two audio interfaces first stereo speaker
with 48kHz stereo stream.
Second interface for microphone with 48kHz mono stream.

This example can be used to start working on audio device.
It can be also used to verify ISO endpoints for boards.

Speaker adaptive clock (bound to SOF).
Microphone for now has asynchronous clock.

Volume and mute control while present are not used for data stream
modification.

**Additional context**
To test:
build for board that have ISO support (currently **NRF52**, **DA1469x**, **STM32F4** (synopsys based)).
Select speaker in system settings (Windows or Linux) play audio (i.e. from youtube).
Use *audacity* to record sound from headset.
LED blinks very fast when data is streamed from HOST (after set interface on speaker is called with
alt settings set to 1)

Know issues:
- ~**NRF52** is not building due to missing **core_cm4.h** (other examples probably will also not build for Nordic boards).~ not an issue.
- **synopsys** has hard coded receive FIFO depth set to 64 bytes which is not enough to run this example (196 ISO OUT endpoint size).
@PanRe version handled this case but I think it was reverted in final merge.

### CODE REVIEW

**Comment by hathach** [comment]
brilliant !!! Thank you very much for the PR, I will check + test it out asap :+1: :+1:

File: examples/device/uac2_headset/src/usb_descriptors.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    /* Standard AC Interface Descriptor(4.7.1) */\
+    TUD_AUDIO_DESC_STD_AC(/*_itfnum*/ ITF_NUM_AUDIO_CONTROL, /*_nEPs*/ 0x00, /*_stridx*/ _stridx),\
+    /* Class-Specific AC Interface Header Descriptor(4.7.2) */\
+    TUD_AUDIO_DESC_CS_AC(/*_bcdADC*/ 0x0200, /*_category*/ AUDIO_FUNC_HEADSET, /*_totallen*/ TUD_AUDIO_DESC_CLK_SRC_LEN+TUD_AUDIO_DESC_FEATURE_UNIT_TWO_CHANNEL_LEN+TUD_AUDIO_DESC_INPUT_TERM_LEN+TUD_AUDIO_DESC_OUTPUT_TERM_LEN+TUD_AUDIO_DESC_INPUT_TERM_LEN+TUD_AUDIO_DESC_OUTPUT_TERM_LEN, /*_ctrl*/ AUDIO_CS_AS_INTERFACE_CTRL_LATENCY_POS),\
+    /* Clock Source Descriptor(4.7.2.1) */\
+    TUD_AUDIO_DESC_CLK_SRC(/*_clkid*/ UAC2_ENTITY_CLOCK, /*_attr*/ 3, /*_ctrl*/ 5, /*_assocTerm*/ 0x00,  /*_stridx*/ 0x00),    \
+    /* Input Terminal Descriptor(4.7.2.4) */\
+    TUD_AUDIO_DESC_INPUT_TERM(/*_termid*/ UAC2_ENTITY_SPK_INPTU_TERMINAL, /*_termtype*/ AUDIO_TERM_TYPE_USB_STREAMING, /*_assocTerm*/ 0x00, /*_clkid*/ UAC2_ENTITY_CLOCK, /*_nchannelslogical*/ 0x02, /*_channelcfg*/ AUDIO_CHANNEL_CONFIG_NON_PREDEFINED, /*_idxchannelnames*/ 0x00, /*_ctrl*/ 0 * (AUDIO_CTRL_R << AUDIO_IN_TERM_CTRL_CONNECTOR_POS), /*_stridx*/ 0x00),\
```
Comment (chegewara) [inline examples/device/uac2_headset/src/usb_descriptors.h:82]: Typo?
`UAC2_ENTITY_SPK_INPTU_TERMINAL`

**Comment by hathach** [comment]
> **Describe the PR**
> This example code creates USB Audio 2.0 headset device.
> Device has two audio interfaces first stereo speaker
> with 48kHz stereo stream.
> Second interface for microphone with 48kHz mono stream.
> 
> This example can be used to start working on audio device.
> It can be also used to verify ISO endpoints for boards.
> 
> Speaker adaptive clock (bound to SOF).
> Microphone for now has asynchronous clock.
> 
> Volume and mute control while present are not used for data stream
> modification.
> 
> **Additional context**
> To test:
> build for board that have ISO support (currently **NRF52**, **DA1469x**, **STM32F4** (synopsys based)).
> Select speaker in system settings (Windows or Linux) play audio (i.e. from youtube).
> Use _audacity_ to record sound from headset.
> LED blinks very fast when data is streamed from HOST (after set interface on speaker is called with
> alt settings set to 1)

This example is incredible, really useful for doing further UAC2 in the future. I didn't test much, but device is recognized and led is blinking fast under Linux
image

> Know issues:
> 
> * **NRF52** is not building due to missing **core_cm4.h** (other examples probably will also not build for Nordic boards).

I actually test with nRF52, there is the CMSIS v4 submodule (will also add v5 later on), please try to run `git submodule update --init` again to see if that work
https://github.com/ARM-software/CMSIS/tree/f2cad4345783c948ed4a7f5cdb02cdc0856366f1

> * **synopsys** has hard coded receive FIFO depth set to 64 bytes which is not enough to run this example (196 ISO OUT endpoint size).
>   @PanRe version handled this case but I think it was reverted in final merge.

ah yes, it is indeed reverted in the merge, since I opt to have an mechanism for USBD to manage this instead of having DCD to handle it all. And yet, I haven't got time to work on that :disappointed: :disappointed: . I did notice the issue with this with the FIXME here https://github.com/hathach/tinyusb/blob/master/src/portable/st/synopsys/dcd_synopsys.c#L216 

For a quick quick walkaround, we can increase the FIFO OUT to let's say 256 when AUDIO class is enabled. Just add TODO/FIXME with some note so that we could address it better later on

File: examples/device/uac2_headset/src/tusb_config.h [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+#error CFG_TUSB_MCU must be defined
+#endif
+
+#if CFG_TUSB_MCU == OPT_MCU_LPC43XX || CFG_TUSB_MCU == OPT_MCU_LPC18XX || CFG_TUSB_MCU == OPT_MCU_MIMXRT10XX
+#define CFG_TUSB_RHPORT0_MODE       (OPT_MODE_DEVICE | OPT_MODE_HIGH_SPEED)
+#else
+#define CFG_TUSB_RHPORT0_MODE       OPT_MODE_DEVICE
+#endif
+
+#define CFG_TUSB_DEBUG              0
```
Comment (hathach) [inline examples/device/uac2_headset/src/tusb_config.h:49]: very minor, would you also wrap the debug in the #ifndef this help when compiling with `LOG=` option for debugging.

File: examples/device/uac2_headset/src/usb_descriptors.h (hunk 1 above)
Comment (kasjer) [reply examples/device/uac2_headset/src/usb_descriptors.h]: thanks, fixed

File: examples/device/uac2_headset/src/tusb_config.h (hunk 2 above)
Comment (kasjer) [reply examples/device/uac2_headset/src/tusb_config.h]: done

**Comment by kasjer** [comment]
@hathach I will leave fixing FIFO issue for synopsys to you. I'm sure you have better overall vision on how it should be done.
It is not part of the headset code anyway so separate PR addressing this will have chance for better review.

**Comment by hathach** [comment]
> @hathach I will leave fixing FIFO issue for synopsys to you. I'm sure you have better overall vision on how it should be done.
> It is not part of the headset code anyway so separate PR addressing this will have chance for better review.

yeah, that makes lots of sense, I will merge this and make an issue for this. Will try to find time this weekend for a better fix.

**Review by hathach — APPROVED** [review]
Thank you very much for this incredible PR. This is really a great working base for us to add more work on UAC2. I really need further reading the UAC2 and more hand-on testing audio class. 

The current audio API is a bit technical that needs lots of reading to get it work. I am hoping to simplify it more to make it easier for user. That is for sure more complicate and need a series of following-up PRs.

**Pull request merged**